File: packages/sdk-typescript/src/resources/team.ts
// V-298c / V-309e — Team RBAC resource.
//
// All six /v1/team/* endpoints, plus the two /v1/teams team-record endpoints. Team membership IS honored on the auth
// path: send `X-Driftstack-Account: acc_<owner-uuid>` to act on the
// resources of an owner you are a member of. The request is authorized
// against your membership role ('admin' or 'member') and the route's
// required scope; without the header every call acts on your own account.

import type { HttpClient } from '../http.js';

export type TeamRole = 'member' | 'admin';

export interface TeamMember {
  id: string;
  owner_account_id: string;
  member_account_id: string;
  member_email: string;
  role: TeamRole;
  invited_at: string;
  accepted_at: string;
  invited_by_account_id: string | null;
}

export interface TeamInvite {
  id: string;
  owner_account_id: string;
  invitee_email: string;
  role: TeamRole;
  expires_at: string;
  invited_by_account_id: string | null;
  accepted_at: string | null;
  created_at: string;
}

export interface TeamOwner {
  owner_account_id: string;
  owner_email: string;
  owner_name: string | null;
  role: TeamRole;
  membership_id: string;
}

export interface TeamMembersList {
  data: TeamMember[];
}

export interface TeamInvitesList {
  data: TeamInvite[];
}

export interface TeamOwnersList {
  data: TeamOwner[];
}

export interface AcceptInviteResponse {
  membership: TeamMember;
}

/**
 * A team as a record — what `GET /v1/teams` returns.
 *
 * Distinct from `TeamOwner`, which describes a team you BELONG to (it carries
 * your `role` and `membership_id`). This one is the team itself.
 */
export interface TeamRecord {
  id: string;
  name: string;
  /**
   * Always `null` today. Published because it exists on the record, but no
   * endpoint sets one — whether team slugs become public URL components is an
   * open decision. Safe to read; will stay null until that is settled.
   */
  slug: string | null;
  owner_account_id: string;
  created_at: string;
  updated_at: string;
}

export interface TeamRecordsList {
  data: TeamRecord[];
}

export interface InviteOptions {
  role?: TeamRole;
}

export class TeamResource {
  constructor(private readonly http: HttpClient) {}

  // Requires broad read or account_owner.
  /** List the teams the calling account OWNS. */
  listTeams(): Promise<TeamRecordsList> {
    return this.http.request<TeamRecordsList>({
      method: 'GET',
      path: '/v1/teams',
    });
  }

  /**
   * Rename a team the calling account owns. account_owner scope required.
   *
   * A 404 covers both an unknown id and a team owned by someone else, and does
   * so deliberately — distinguishing them would let a caller enumerate which
   * team ids exist.
   */
  renameTeam(teamId: string, name: string): Promise<{ team: TeamRecord }> {
    return this.http.request<{ team: TeamRecord }>({
      method: 'PATCH',
      path: `/v1/teams/${encodeURIComponent(teamId)}`,
      body: { name },
    });
  }

  /** Invite an email to the calling owner's team. account_owner scope required. */
  invite(email: string, options: InviteOptions = {}): Promise<{ message: string }> {
    return this.http.request<{ message: string }>({
      method: 'POST',
      path: '/v1/team/invites',
      body: { email, ...(options.role !== undefined ? { role: options.role } : {}) },
    });
  }

  // Requires broad read or account_owner.
  /** List confirmed team members for the calling owner. */
  listMembers(): Promise<TeamMembersList> {
    return this.http.request<TeamMembersList>({
      method: 'GET',
      path: '/v1/team/members',
    });
  }

  // Requires broad read or account_owner.
  /** List pending (unaccepted, unexpired) invites for the calling owner. */
  listInvites(): Promise<TeamInvitesList> {
    return this.http.request<TeamInvitesList>({
      method: 'GET',
      path: '/v1/team/invites',
    });
  }

  // Requires broad read or account_owner.
  /** List owner workspaces the calling account has joined. */
  listOwners(): Promise<TeamOwnersList> {
    return this.http.request<TeamOwnersList>({
      method: 'GET',
      path: '/v1/team/owners',
    });
  }

  /**
   * Accept a pending invite. The accepting account's email MUST match
   * the invitee email — server enforces; mismatched accept returns 409.
   * Requires account_owner (dashboard web sessions satisfy it).
   */
  acceptInvite(token: string): Promise<AcceptInviteResponse> {
    return this.http.request<AcceptInviteResponse>({
      method: 'POST',
      path: '/v1/team/invites/accept',
      body: { token },
    });
  }

  /** Remove a member by membership id. account_owner scope required. */
  removeMember(membershipId: string): Promise<void> {
    return this.http.request<void>({
      method: 'DELETE',
      path: `/v1/team/members/${encodeURIComponent(membershipId)}`,
    });
  }
}

File: packages/sdk-python/src/driftstack/resources/team.py
"""V-298c / V-309f — Team RBAC resource.

All six /v1/team/* endpoints, plus the two /v1/teams team-record endpoints. Team membership IS honored on the auth
path: send ``X-Driftstack-Account: acc_<owner-uuid>`` to act on the
resources of an owner you are a member of. The request is authorized
against your membership role (``admin`` or ``member``) and the route's
required scope; without the header every call acts on your own account.
"""

from __future__ import annotations

from typing import Any, Literal
from urllib.parse import quote

from pydantic import BaseModel

from driftstack.http import AsyncHttpClient, HttpClient, parse_model

TeamRole = Literal["member", "admin"]


class TeamMember(BaseModel):
    id: str
    owner_account_id: str
    member_account_id: str
    member_email: str
    role: TeamRole
    invited_at: str
    accepted_at: str
    invited_by_account_id: str | None


class TeamInvite(BaseModel):
    id: str
    owner_account_id: str
    invitee_email: str
    role: TeamRole
    expires_at: str
    invited_by_account_id: str | None
    accepted_at: str | None
    created_at: str


class TeamOwner(BaseModel):
    owner_account_id: str
    owner_email: str
    owner_name: str | None
    role: TeamRole
    membership_id: str


class TeamMembersList(BaseModel):
    data: list[TeamMember]


class TeamInvitesList(BaseModel):
    data: list[TeamInvite]


class TeamOwnersList(BaseModel):
    data: list[TeamOwner]


class AcceptInviteResponse(BaseModel):
    membership: TeamMember


class TeamRecord(BaseModel):
    """A team as a record — what ``GET /v1/teams`` returns.

    Distinct from ``TeamOwner``, which describes a team you BELONG to (it
    carries your ``role`` and ``membership_id``). This one is the team itself.

    ``slug`` is always ``None`` today: the field exists on the record but no
    endpoint sets one, because whether team slugs become public URL components
    is an open decision. Safe to read; it stays ``None`` until that is settled.
    """

    id: str
    name: str
    slug: str | None
    owner_account_id: str
    created_at: str
    updated_at: str


class TeamRecordsList(BaseModel):
    data: list[TeamRecord]


class RenameTeamResponse(BaseModel):
    team: TeamRecord


class TeamResource:
    """Synchronous team resource."""

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def invite(self, email: str, *, role: TeamRole | None = None) -> dict[str, Any]:
        body: dict[str, Any] = {"email": email}
        if role is not None:
            body["role"] = role
        data = self._http.request("POST", "/v1/team/invites", json_body=body)
        return data  # {"message": ...}

    # Requires broad read or account_owner.
    def list_members(self) -> TeamMembersList:
        data = self._http.request("GET", "/v1/team/members")
        return parse_model(TeamMembersList, data)

    # Requires broad read or account_owner.
    def list_invites(self) -> TeamInvitesList:
        data = self._http.request("GET", "/v1/team/invites")
        return parse_model(TeamInvitesList, data)

    # Requires broad read or account_owner.
    def list_owners(self) -> TeamOwnersList:
        data = self._http.request("GET", "/v1/team/owners")
        return parse_model(TeamOwnersList, data)

    # Requires account_owner.
    def accept_invite(self, token: str) -> AcceptInviteResponse:
        data = self._http.request("POST", "/v1/team/invites/accept", json_body={"token": token})
        return parse_model(AcceptInviteResponse, data)

    def remove_member(self, membership_id: str) -> None:
        self._http.request("DELETE", f"/v1/team/members/{quote(membership_id, safe='')}")

    # Requires broad read or account_owner.
    def list_teams(self) -> TeamRecordsList:
        data = self._http.request("GET", "/v1/teams")
        return parse_model(TeamRecordsList, data)

    # Requires account_owner. 404 covers both an unknown id and a team owned by
    # someone else, deliberately — telling them apart would let a caller
    # enumerate which team ids exist.
    def rename_team(self, team_id: str, name: str) -> RenameTeamResponse:
        data = self._http.request(
            "PATCH", f"/v1/teams/{quote(team_id, safe='')}", json_body={"name": name}
        )
        return parse_model(RenameTeamResponse, data)


class AsyncTeamResource:
    """Async team resource."""

    def __init__(self, http: AsyncHttpClient) -> None:
        self._http = http

    async def invite(self, email: str, *, role: TeamRole | None = None) -> dict[str, Any]:
        body: dict[str, Any] = {"email": email}
        if role is not None:
            body["role"] = role
        data = await self._http.request("POST", "/v1/team/invites", json_body=body)
        return data

    # Requires broad read or account_owner.
    async def list_members(self) -> TeamMembersList:
        data = await self._http.request("GET", "/v1/team/members")
        return parse_model(TeamMembersList, data)

    # Requires broad read or account_owner.
    async def list_invites(self) -> TeamInvitesList:
        data = await self._http.request("GET", "/v1/team/invites")
        return parse_model(TeamInvitesList, data)

    # Requires broad read or account_owner.
    async def list_owners(self) -> TeamOwnersList:
        data = await self._http.request("GET", "/v1/team/owners")
        return parse_model(TeamOwnersList, data)

    # Requires account_owner.
    async def accept_invite(self, token: str) -> AcceptInviteResponse:
        data = await self._http.request(
            "POST", "/v1/team/invites/accept", json_body={"token": token}
        )
        return parse_model(AcceptInviteResponse, data)

    async def remove_member(self, membership_id: str) -> None:
        await self._http.request("DELETE", f"/v1/team/members/{quote(membership_id, safe='')}")

    # Requires broad read or account_owner.
    async def list_teams(self) -> TeamRecordsList:
        data = await self._http.request("GET", "/v1/teams")
        return parse_model(TeamRecordsList, data)

    # Requires account_owner. 404 covers both an unknown id and a team owned by
    # someone else, deliberately — telling them apart would let a caller
    # enumerate which team ids exist.
    async def rename_team(self, team_id: str, name: str) -> RenameTeamResponse:
        data = await self._http.request(
            "PATCH", f"/v1/teams/{quote(team_id, safe='')}", json_body={"name": name}
        )
        return parse_model(RenameTeamResponse, data)
